chalk-ir/src/visit/boring_impls.rs

```diff
@@ -0,0 +1,341 @@
+        for e in self {
+            result = result.and_then(|| e.visit_with(visitor, outer_binder))
+        }
+        result
+    }
+}
+
+impl<T: Visit<I>, I: Interner> Visit<I> for Box<T> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        (**self).visit_with(visitor, outer_binder)
+    }
+}
+
+impl<T: Visit<I>, I: Interner> Visit<I> for Arc<T> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        (**self).visit_with(visitor, outer_binder)
+    }
+}
+
+macro_rules! tuple_visit {
+    ($($n:ident),*) => {
+        impl<$($n: Visit<I>,)* I: Interner> Visit<I> for ($($n,)*) {
+            fn visit_with<'i, R: VisitResult>(&self, visitor: &mut dyn Visitor<'i, I, Result = R>, outer_binder: DebruijnIndex) -> R where I: 'i
+            {
+                #[allow(non_snake_case)]
+                let &($(ref $n),*) = self;
+                R::new()
+                  $(.and_then(|| $n.visit_with(visitor, outer_binder)))*
+            }
+        }
+    }
+}
+
+tuple_visit!(A, B);
+tuple_visit!(A, B, C);
+tuple_visit!(A, B, C, D);
+tuple_visit!(A, B, C, D, E);
+
+impl<T: Visit<I>, I: Interner> Visit<I> for Option<T> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        match self {
+            Some(e) => e.visit_with(visitor, outer_binder),
+            None => R::new(),
+        }
+    }
+}
+
+impl<I: Interner> Visit<I> for Parameter<I> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let interner = visitor.interner();
+        self.data(interner).visit_with(visitor, outer_binder)
+    }
+}
+
+impl<I: Interner> Visit<I> for Substitution<I> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let interner = visitor.interner();
+        let mut result = R::new();
+
+        for p in self.iter(interner) {
+            result = result.and_then(|| p.visit_with(visitor, outer_binder));
+        }
+
+        result
+    }
+}
+
+impl<I: Interner> Visit<I> for Goals<I> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let interner = visitor.interner();
+        let mut result = R::new();
+
+        for p in self.iter(interner) {
```

Comment:
same as the for loop above; we should probably make a helper for `for` loop iterator regardless